### PR TITLE
Keep precise argument sizes between import and morph.

### PR DESCRIPTION
**Pull request by sandreenko**
On arm64 OSX we need to know precise argument sizes. I have added support for that between morph and lowering in #42503, this PR supports in between importation and morph.

When do we need it:
```
IL: load byte; call1(int) -> IR: CALL1(IND byte);
IL load int; call2(byte) -> IR: CALL2(IND int);
```
without the change, morph was seeing the incorrect type of the nodes but for other platforms it was not important (because all arguments were passed in TARGET_POINTER_SIZE slots).

I have stopped on the solution that adds a special GT_OPER to keep this information between these phases and a debug code in morph to check that it has survived phases between them (mostly inlining). Thanks @CarolEidt for the idea.
I want to enable this code for all platforms because:
1. do not want to introduce platform dependency in HIR;
2. it will make arm64 OSX work cheaper because we will catch bugs in this component on other platforms until we have arm64 OSX in ci;
4. the overhead is small, it should not affect TP;

I was also considering other solutions, the main 2:
1. add a new artificial structure to keep this information, map/dictionary, but
- we can't increase call node size;
- it will be an error-prone component from our previous experience with `fgArgInfo`;
- it will have non-const access time or memory overhead, also non-trivial copying;

2. change `impImplicitIorI4Cast` to insert cast not only between i4 and i8 but between all sizes, but
- we don't need it if it is not under a call;
- it will require changes during inlining to ignore such nodes;
- it will be easy for such nodes to slip through morph phase;

~~No asm diffs (SPMI x64/x86).~~  Update: ci shows related failures during the check phase, I will fix them soon, but I think it should not block review.

### CODE REVIEW

File: src/coreclr/src/jit/importer.cpp
Comment (sandreenko) [inline src/coreclr/src/jit/importer.cpp:965]: I wonder why we did not see a warning about an assignment inside `if`.

File: src/coreclr/src/jit/morph.cpp
Comment (sandreenko) [inline src/coreclr/src/jit/morph.cpp:2899]: `callSig == nullptr` means mostly a helper call, all helper calls that I have checked are working with i4 and i8 types but I expect I will have to add more `PUTARG_TYPE` there once I run more tests on arm64 OSX and find bugs with missed type information.

**Comment by sandreenko** [comment]
PTAL @CarolEidt @AndyAyersMS @dotnet/jit-contrib

**Comment by AndyAyersMS** [comment]
I'm puzzled why we need to do things this way -- doesn't the callee sig tell us what size the callee expects?

**Comment by sandreenko** [comment]
> I'm puzzled why we need to do things this way -- doesn't the callee sig tell us what size the callee expects?

We get this information from callee sig, but it is available only during importation and then we lose it. In debug we save it in 
```
#ifdef DEBUG
    // Used to register callsites with the EE
    CORINFO_SIG_INFO* callSig;
#endif
```
if we decide to keep it in Release it will increase the size of all large nodes.

**Comment by AndyAyersMS** [comment]
Seems like we could deal with that easily enough, we have various extensions already hanging off of calls. For instance we could refactor `TailCallSiteInfo`  which already holds onto a `CORINFO_SIG_INFO` and use it more broadly for cases where we need the sig later in compilation.

Is there other benefit to having these extra IR nodes beyond being able to handle this ABI quirk?

**Comment by sandreenko** [comment]
> Is there other benefit to having these extra IR nodes beyond being able to handle this ABI quirk?

These extra nodes are only for the ABI quirk, I do not see benefits to other places from it. I was thinking to postpone casts creations but it was a long shot (if we pass 1 byte as 1 byte we don't need `CAST int<-byte<-int`).
I was using the signature in my prototype (from the debug field) to get the correct handle and there were some issues:
1. no way to handle calls without a signature if we get such;
2. not easy to match the signature to the actual argument nodes in morph.  I am trying to do it for checking now, but with additional arguments from both sides of the list it is not trivial;
2.  need to do extra calls to `strip(info.compCompHnd->getArgType(sig, sigArgs, &classHnd))` (minor);

File: src/coreclr/src/jit/flowgraph.cpp
Comment (CarolEidt) [inline src/coreclr/src/jit/flowgraph.cpp:22774]: Prior to this comment, neither the function header nor the code above talk about swapping the type. It would be good to describe why, and in what scenarios, such type swapping is done (including why `retType` may not be the same as `newType`).

File: src/coreclr/src/jit/gentree.h
Comment (CarolEidt) [inline src/coreclr/src/jit/gentree.h:7078]: I'm curious why you added this here - that is, I can see why we wouldn't expect to encounter these under a `COMMA`, but I'm not sure why this special assert is needed.

File: src/coreclr/src/jit/importer.cpp
Comment (CarolEidt) [inline src/coreclr/src/jit/importer.cpp:990]: This assert is redundant.

File: src/coreclr/src/jit/importer.cpp
Comment (CarolEidt) [inline src/coreclr/src/jit/importer.cpp:19028]: Would it make sense to add something like `gtSkipPutArgType()` (similar to `gtEffectiveVal()`)?

File: src/coreclr/src/jit/morph.cpp
Comment (CarolEidt) [inline src/coreclr/src/jit/morph.cpp:2894]:
```suggestion
    // Check that we have valid information about call's argument types.
```

File: src/coreclr/src/jit/morph.cpp
Comment (CarolEidt) [inline src/coreclr/src/jit/morph.cpp:2917]: Should this be `!=`? Otherwise, the loop below will never execute, right?

File: src/coreclr/src/jit/rationalize.cpp
Comment (CarolEidt) [inline src/coreclr/src/jit/rationalize.cpp:306]:
```suggestion
                // QMARK and PUT_ARG_TYPE nodes should have been removed before this phase.
```

**Review by CarolEidt — COMMENTED** [review]
Some initial comments and suggestions

File: src/coreclr/src/jit/gentree.h
Comment (sandreenko) [reply src/coreclr/src/jit/gentree.h:7078]: We have several similar functions like `gtEffectiveVal`, `gtRetExprVal` and `gtSkipPutArgType`. When we call each of them we often don't expect nodes that can be skipped by two others to be on top.
For example, when we call `gtEffectiveVal` we don't expect something like `GT_RET_EXPR(GT_COMMA)` or `GT_PUTARG_TYPE(GT_COMMA)`, similar for `gtRetExprVal` it should not see `GT_PUTARG_TYPE(GT_PUTARG_TYPE)`.
If we do it means the caller did not check for these nodes and we don't skip what we hoped to skip, it doesn't cause correctness issues in my understanding but introduces asm regressions when we can't parse something like we did before the change.

File: src/coreclr/src/jit/morph.cpp
Comment (sandreenko) [reply src/coreclr/src/jit/morph.cpp:2917]: It is a temporary condition, the idea here is to match the signature with the call arguments as we have them in morph and check that we still have the correct types. However, it is tricky to do because we could have some nodes added to `gtCallArgs` between importation and call to this function (also this function can be called several times). 
So we need to skip all additional nodes that were added and I have problems with identifying them. I am planning to fix it once I pass all the other tests.

File: src/coreclr/src/jit/flowgraph.cpp
Comment (sandreenko) [inline src/coreclr/src/jit/flowgraph.cpp:23852]: This new condition allows us to always the argument as a temp with the correct type, we need it because for a correct direct substitution we need to know if the parent is a call or not but we don't have this information.
We create the node when we see an IL opcode like `ldlocal` and don't have parent information, later, when the node is used, we can't distinguish it from a regular `LCL_VAR`.
The change causes a small regression on all platforms:
```
windows x64 crossgen: 5134 (0.02% of base)
windows x64 pmi:  14911 (0.03% of base)
windows arm64 crossgen: 5688 (0.01% of base)
linux x64 crossgen: 9036 (0.02% of base)

```
, etc. all lower than 0.04%.

What happens there is:
1. during importation we create a `LCL_VAR tempForOurArg`, save a pointer to it;
2. in `fgInsertInlineeBlocks` we iterate over all such pointers and do in-place replacement if there was only one use of the lclVar, but we don't know the user, so we can't say if it is a call that need a `PUTARG` or not.

I have not found a simple way to avoid it or compensate for it.

**Review by sandreenko — COMMENTED** [review]
The PR was updated and is passing the tests now.
Thanks to @AndyAyersMS for helping me offline with the previous test failures. 

The changes include #44790 that I hope to merge first.

This is the last big JIT change for arm64 apple, with it we will have precise arguments sizes from importer to codegen. The next changes will be on VM side for interop stubs.

**Review by AndyAyersMS — COMMENTED** [review]
Left you some questions about this change.

File: src/coreclr/src/jit/flowgraph.cpp
Comment (sandreenko) [reply src/coreclr/src/jit/flowgraph.cpp:5182]: I do not need this for correctness.  I am planning to merge them in https://github.com/dotnet/runtime/pull/44790#issuecomment-729259087, if we decide not to merge it I will need to change this code to use `argNode->gtSkipPutArgType()->gtRetExprVal()`.

File: src/coreclr/src/jit/compiler.h
Comment (sandreenko) [reply src/coreclr/src/jit/compiler.h:4362]: > I don't understand why you removed the bbFlags here, I thought @erozenfeld had iterated a few times to get these properly updated.

Now we save `RET_EXPR` in  `inlCurArgInfo->argNode` so we can get the flags when we need them from `inlCurArgInfo->argNode`.

I will start stress testing.

Comment (erozenfeld) [reply src/coreclr/src/jit/compiler.h:4362]: Please make sure you run libraries stress tests. I believe most of the failures I fixed were from those runs.

Comment (AndyAyersMS) [reply src/coreclr/src/jit/compiler.h:4362]: Note these are BB flags, not tree node flags.

File: src/coreclr/src/jit/flowgraph.cpp
Comment (sandreenko) [inline src/coreclr/src/jit/flowgraph.cpp:23831]: > See note above; I don't believe you can set these to zero like this and pass jit stress.

we read the flags here exactly how we read them in `impInlineInitVars` before but maybe I am missing something, will check.

File: src/coreclr/src/jit/compiler.h
Comment (sandreenko) [reply src/coreclr/src/jit/compiler.h:4362]: the library test have passed, the only failure is due to #42719

File: src/coreclr/src/jit/morph.cpp
Comment (sandreenko) [reply src/coreclr/src/jit/morph.cpp:18410]: Missed that, thanks

File: src/coreclr/src/jit/flowgraph.cpp
Comment (AndyAyersMS) [reply src/coreclr/src/jit/flowgraph.cpp:23834]: I think your changes are ok. Sorry for the false alarm.

Looks like the flags are captured when the `gtRetExpr` is created, and so by holding onto that node instead of tunneling through it, you no longer need to pass the flags around.

File: src/coreclr/src/jit/flowgraph.cpp
Comment (AndyAyersMS) [reply src/coreclr/src/jit/flowgraph.cpp]: I just approved #44790.

**Review by AndyAyersMS — APPROVED** [review]
I think this is good to merge once you've updated for #44790 and perhaps added a comment indicating why `PUTARG_TYPE` can't be directly substituted.

**Comment by sandreenko** [comment]
I have checked that the diffs are still as expected before the merge, PR and jitstress are clear,
the only ci failure is due to https://github.com/dotnet/runtime/issues/45168.

**Comment by sandreenko** [comment]
Thanks @AndyAyersMS and @CarolEidt for your help with this change.

**Pull request merged**
